cmd/promtool/unittest.go

```diff
@@ -0,0 +1,507 @@
+		if ers != nil {
+			errs = append(errs, ers...)
+		}
+	}
+
+	if len(errs) > 0 {
+		return errs
+	}
+	return nil
+}
+
+// unitTestFile holds the contents of a single unit test file.
+type unitTestFile struct {
+	RuleFiles          []string      `yaml:"rule_files"`
+	EvaluationInterval time.Duration `yaml:"evaluation_interval,omitempty"`
+	GroupEvalOrder     []string      `yaml:"group_eval_order"`
+	Tests              []testGroup   `yaml:"tests"`
+}
+
+func (utf *unitTestFile) maxEvalTime() time.Duration {
+	var maxd time.Duration
+	for _, t := range utf.Tests {
+		d := t.maxEvalTime()
+		if d > maxd {
+			maxd = d
+		}
+	}
+	return maxd
+}
+
+// testGroup is a group of input series and tests associated with it.
+type testGroup struct {
+	Interval       time.Duration    `yaml:"interval"`
+	InputSeries    []series         `yaml:"input_series"`
+	AlertRuleTest  []alertTestCase  `yaml:"alert_rule_test,omitempty"`
+	PromqlExprTest []promqlTestCase `yaml:"promql_expr_test,omitempty"`
+}
+
+// test performs the unit tests.
+func (tg *testGroup) test(mint, maxt time.Time, evalInterval time.Duration, groupOrderMap map[string]int, ruleFiles ...string) []error {
+
+	// Setup testing suite.
+	suite, err := promql.NewTest(nil, tg.seriesLoadingString())
+	if err != nil {
+		return []error{err}
+	}
+	defer suite.Close()
+
+	err = suite.Run()
+	if err != nil {
+		return []error{err}
+	}
+
+	// Load the rule files.
+	opts := &rules.ManagerOptions{
+		QueryFunc:  rules.EngineQueryFunc(suite.QueryEngine(), suite.Storage()),
+		Appendable: suite.Storage(),
+		Context:    context.Background(),
+		NotifyFunc: func(ctx context.Context, expr string, alerts ...*rules.Alert) error { return nil },
+	}
+	m := rules.NewManager(opts)
+	groupsMap, ers := m.LoadGroups(tg.Interval, ruleFiles...)
+	if ers != nil {
+		return ers
+	}
+	groups := orderedGroups(groupsMap, groupOrderMap)
+
+	// Pre-processing some data for testing alerts.
+	// All this preparating so that we can test the alerts as and when we evaluate
+	// the rules so that we can avoid storing them in memory, as number of eval might
+	// get very big.
+
+	// All the `eval_time` for which we have unit tests.
+	var alertEvalTimes []time.Duration
+	// Map of all the eval_time+alertname combination present in the unit tests.
+	presentInTest := make(map[time.Duration]map[string]struct{})
+	// Map of all the unit tests for given eval_time.
+	alertTests := make(map[time.Duration][]alertTestCase)
+	for _, art := range tg.AlertRuleTest {
+		pos := func() int {
+			for i, d := range alertEvalTimes {
+				if d > art.EvalTime {
+					return i
+				}
+			}
+			return len(alertEvalTimes)
+		}()
+		alertEvalTimes = append(alertEvalTimes[:pos], append([]time.Duration{art.EvalTime}, alertEvalTimes[pos:]...)...)
+
+		if _, ok := presentInTest[art.EvalTime]; !ok {
+			presentInTest[art.EvalTime] = make(map[string]struct{})
+		}
+		presentInTest[art.EvalTime][art.Alertname] = struct{}{}
+
+		if _, ok := alertTests[art.EvalTime]; ok {
```

Comment:
Could just be: `alertTests[art.EvalTime] = append(alertTests[art.EvalTime], art)`